From ec6ab34e9588fd28ec4e938062acfff1ada8fc8f Mon Sep 17 00:00:00 2001
From: Aihua Xu <aihuaxu@gmail.com>
Date: Thu, 30 Jul 2015 09:38:31 -0700
Subject: [PATCH 0216/1164] HIVE-11258 - The function drop_database_core() of
 HiveMetaStore.java may not drop all the tables
 (Aihua Xu, reviewed by Chao Sun)

---
 .../hadoop/hive/metastore/HiveMetaStore.java       |   11 ++++-------
 1 file changed, 4 insertions(+), 7 deletions(-)

diff --git a/metastore/src/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java b/metastore/src/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java
index 6eee868..ed88c8e 100644
--- a/metastore/src/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java
+++ b/metastore/src/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java
@@ -1025,14 +1025,9 @@ private void drop_database_core(RawStore ms,
             ConfVars.METASTORE_BATCH_RETRIEVE_MAX);
 
         int startIndex = 0;
-        int endIndex = -1;
         // retrieve the tables from the metastore in batches to alleviate memory constraints
-        while (endIndex < allTables.size() - 1) {
-          startIndex = endIndex + 1;
-          endIndex = endIndex + tableBatchSize;
-          if (endIndex >= allTables.size()) {
-            endIndex = allTables.size() - 1;
-          }
+        while (startIndex < allTables.size()) {
+          int endIndex = Math.min(startIndex + tableBatchSize, allTables.size());
 
           List<Table> tables = null;
           try {
@@ -1068,6 +1063,8 @@ private void drop_database_core(RawStore ms,
               // Drop the table but not its data
               drop_table(name, table.getTableName(), false);
             }
+
+            startIndex = endIndex;
           }
         }
 
-- 
1.7.9.5

